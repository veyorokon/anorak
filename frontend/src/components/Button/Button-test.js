/**
 * Copyright IBM Corp. 2016, 2018
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import { iconSearch } from 'carbon-icons';
import Search16 from '@carbon/icons-react/lib/search/16';
import Button from '../Button';
import Link from '../Link';
import ButtonSkeleton from '../Button/Button.Skeleton';
import { shallow, mount } from 'enzyme';
import { componentsX } from '../../internal/FeatureFlags';

describe('Button', () => {
  describe('Renders common props as expected', () => {
    const wrapper = shallow(
      // eslint-disable-next-line jsx-a11y/tabindex-no-positive
      <Button tabIndex={2} className="extra-class">
        <div className="child">child</div>
        <div className="child">child</div>
      </Button>
    );

    const wrapperHref = shallow(
      // eslint-disable-next-line jsx-a11y/tabindex-no-positive
      <Button tabIndex={2} className="extra-class" href="/home">
        <div className="child">child</div>
        <div className="child">child</div>
      </Button>
    );

    it('Renders children as expected', () => {
      expect(wrapper.find('.child').length).toBe(2);
      expect(wrapperHref.find('.child').length).toBe(2);
    });

    it('Should set tabIndex if one is passed via props', () => {
      expect(wrapper.props().tabIndex).toEqual(2);
      expect(wrapperHref.props().tabIndex).toEqual(2);
    });

    it('Should add extra classes via className', () => {
      expect(wrapper.hasClass('extra-class')).toBe(true);
      expect(wrapperHref.hasClass('extra-class')).toBe(true);
    });
  });

  describe('Renders <button> props as expected', () => {
    const wrapper = shallow(
      // eslint-disable-next-line jsx-a11y/tabindex-no-positive
      <Button tabIndex={2}>
        <div className="child">child</div>
        <div className="child">child</div>
      </Button>
    );

    it('Renders as a <button> element without an href', () => {
      expect(wrapper.is('button')).toBe(true);
    });

    it('Should set disabled to false by default', () => {
      expect(wrapper.props().disabled).toBe(false);
    });

    it('Should set disabled if one is passed via props', () => {
      wrapper.setProps({ disabled: true });
      expect(wrapper.props().disabled).toBe(true);
    });

    it('Should set type to button by default', () => {
      expect(wrapper.props().type).toEqual('button');
    });

    it('Should only set type to [button, reset or submit] if one is passed via props', () => {
      wrapper.setProps({ type: 'reset' });
      expect(wrapper.props().type).toEqual('reset');
      wrapper.setProps({ type: 'submit' });
      expect(wrapper.props().type).toEqual('submit');
    });
  });

  describe('Renders <a> props as expected', () => {
    const wrapper = shallow(
      // eslint-disable-next-line jsx-a11y/tabindex-no-positive
      <Button href="#" tabIndex={2}>
        <div className="child">child</div>
        <div className="child">child</div>
      </Button>
    );

    it('Renders as an <a> element with an href', () => {
      expect(wrapper.is('a')).toBe(true);
    });

    it('Should always render with [role="button"] by default', () => {
      expect(wrapper.props().role).toEqual('button');
    });
  });

  describe('Renders arbitrary component with correct props', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = shallow(
        <Button as={Link} data-foo="foo">
          <div className="child">child</div>
          <div className="child">child</div>
        </Button>
      );
    });
    it('renders as a Link with data attribute', () => {
      expect(wrapper.is(Link)).toBe(true);
      expect(wrapper.is('[data-foo="foo"]')).toBe(true);
    });
  });

  describe('Renders icon buttons', () => {
    const iconButton = mount(
      <Button
        icon={!componentsX && iconSearch}
        renderIcon={componentsX && Search16}
        iconDescription="Search">
        Search
      </Button>
    );
    const icon = iconButton.find('svg');

    it('should have the appropriate icon', () => {
      expect(icon.hasClass('bx--btn__icon')).toBe(true);
    });

    it('should return error if icon given without description', () => {
      const props = {
        icon: 'search',
      };
      // eslint-disable-next-line quotes
      const error = new Error(
        'icon/renderIcon property specified without also providing an iconDescription property.'
      );
      expect(Button.propTypes.iconDescription(props)).toEqual(error);
    });
  });

  describe('Renders custom icon buttons', () => {
    const iconButton = mount(
      <Button renderIcon={Search16} iconDescription="Search">
        Search
      </Button>
    );
    const originalIcon = mount(<Search16 />).find('svg');
    const icon = iconButton.find('svg');

    it('should have the appropriate icon', () => {
      expect(icon.hasClass('bx--btn__icon')).toBe(true);
      expect(icon.find(':not(svg):not(title)').html()).toBe(
        originalIcon.children().html()
      );
    });

    it('should return error if icon given without description', () => {
      const props = {
        icon: 'search',
      };
      // eslint-disable-next-line quotes
      const error = new Error(
        'icon/renderIcon property specified without also providing an iconDescription property.'
      );
      expect(Button.propTypes.iconDescription(props)).toEqual(error);
    });
  });
});

describe('Primary Button', () => {
  describe('Renders as expected', () => {
    const wrapper = shallow(<Button className="extra-class" />);

    it('Has the expected classes', () => {
      expect(wrapper.hasClass('bx--btn')).toEqual(true);
    });

    it('Should add extra classes that are passed via className', () => {
      expect(wrapper.hasClass('extra-class')).toEqual(true);
    });
  });
});

describe('Secondary Button', () => {
  describe('Renders as expected', () => {
    const wrapper = shallow(
      <Button kind="secondary" className="extra-class" />
    );

    it('Has the expected classes', () => {
      expect(wrapper.hasClass('bx--btn--secondary')).toEqual(true);
    });

    it('Should add extra classes that are passed via className', () => {
      expect(wrapper.hasClass('extra-class')).toEqual(true);
    });
  });
});

describe('Ghost Button', () => {
  describe('Renders as expected', () => {
    const wrapper = shallow(<Button kind="ghost" className="extra-class" />);

    it('Has the expected classes', () => {
      expect(wrapper.hasClass('bx--btn--ghost')).toEqual(true);
    });

    it('Should add extra classes that are passed via className', () => {
      expect(wrapper.hasClass('extra-class')).toEqual(true);
    });
  });
});

describe('Small Button', () => {
  describe('Renders as expected', () => {
    const wrapper = shallow(<Button small className="extra-class" />);

    it('Has the expected classes for small', () => {
      expect(wrapper.hasClass('bx--btn--sm')).toEqual(true);
    });

    it('Should add extra classes that are passed via className', () => {
      expect(wrapper.hasClass('extra-class')).toEqual(true);
    });
  });
});

describe('DangerButton', () => {
  describe('Renders as expected', () => {
    const wrapper = shallow(<Button kind="danger" className="extra-class" />);

    it('Has the expected classes', () => {
      expect(wrapper.hasClass('bx--btn--danger')).toEqual(true);
    });

    it('Should add extra classes that are passed via className', () => {
      expect(wrapper.hasClass('extra-class')).toEqual(true);
    });
  });
});

describe('danger--primaryButton', () => {
  describe('Renders as exptected', () => {
    const wrapper = shallow(
      <Button kind="danger--primary" className="extra-class" />
    );

    it('Has the expected classes', () => {
      expect(wrapper.hasClass('bx--btn--danger--primary')).toEqual(true);
    });

    it('Should add extra classes that are passed via className', () => {
      expect(wrapper.hasClass('extra-class')).toEqual(true);
    });
  });
});

describe('TertiaryButton', () => {
  describe('Renders as exptected', () => {
    const wrapper = shallow(<Button kind="tertiary" className="extra-class" />);

    it('Has the expected classes', () => {
      expect(wrapper.hasClass('bx--btn--tertiary')).toEqual(true);
    });

    it('Should add extra classes that are passed via className', () => {
      expect(wrapper.hasClass('extra-class')).toEqual(true);
    });
  });
});

describe('ButtonSkeleton', () => {
  describe('Renders as expected', () => {
    const wrapper = shallow(<ButtonSkeleton />);

    it('Has the expected classes', () => {
      expect(wrapper.hasClass('bx--skeleton')).toEqual(true);
      expect(wrapper.hasClass('bx--btn')).toEqual(true);
    });
  });

  describe('Renders <a> props as expected', () => {
    const wrapper = shallow(
      // eslint-disable-next-line jsx-a11y/tabindex-no-positive
      <ButtonSkeleton href="#" />
    );

    it('Renders as an <a> element with an href', () => {
      expect(wrapper.is('a')).toBe(true);
    });

    it('Should always render with [role="button"] by default', () => {
      expect(wrapper.props().role).toEqual('button');
    });
  });
});

describe('Small ButtonSkeleton', () => {
  describe('Renders as expected', () => {
    const wrapper = shallow(<ButtonSkeleton small />);

    it('Has the expected classes for small', () => {
      expect(wrapper.hasClass('bx--btn--sm')).toEqual(true);
      expect(wrapper.hasClass('bx--skeleton')).toEqual(true);
    });
  });
});
