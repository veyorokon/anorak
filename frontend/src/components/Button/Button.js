/**
 * Copyright IBM Corp. 2016, 2018
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import PropTypes from 'prop-types';
import React from 'react';
import Icon from '../Icon';
import classNames from 'classnames';
import warning from 'warning';
import { settings } from 'carbon-components';
import { ButtonTypes } from '../../prop-types/types';
import { breakingChangesX } from '../../internal/FeatureFlags';

const { prefix } = settings;

let didWarnAboutDeprecation = false;

const Button = React.forwardRef(
  (
    {
      children,
      as,
      className,
      disabled,
      small,
      kind,
      href,
      tabIndex,
      type,
      inputref,
      renderIcon,
      icon,
      iconDescription,
      ...other
    },
    ref
  ) => {
    const buttonClasses = classNames(className, {
      [`${prefix}--btn`]: true,
      [`${prefix}--btn--sm`]: small,
      [`${prefix}--btn--primary`]: kind === 'primary',
      [`${prefix}--btn--danger`]: kind === 'danger',
      [`${prefix}--btn--secondary`]: kind === 'secondary',
      [`${prefix}--btn--ghost`]: kind === 'ghost',
      [`${prefix}--btn--danger--primary`]: kind === 'danger--primary',
      [`${prefix}--btn--tertiary`]: kind === 'tertiary',
      [`${prefix}--btn--disabled`]: disabled,
    });

    const commonProps = {
      tabIndex,
      className: buttonClasses,
      ref: breakingChangesX ? ref : ref || inputref,
    };

    if (__DEV__ && breakingChangesX && icon) {
      warning(
        didWarnAboutDeprecation,
        'The `icon` property in the `Button` component is being removed in the next release of ' +
          '`carbon-components-react`. Please use `renderIcon` instead.'
      );
      didWarnAboutDeprecation = true;
    }

    const hasRenderIcon = Object(renderIcon) === renderIcon;
    const ButtonImageElement = hasRenderIcon
      ? renderIcon
      : !breakingChangesX && icon && Icon;
    const buttonImage = !ButtonImageElement ? null : (
      <ButtonImageElement
        icon={!hasRenderIcon && Object(icon) === icon ? icon : undefined}
        name={!hasRenderIcon && Object(icon) !== icon ? icon : undefined}
        aria-label={!hasRenderIcon ? undefined : iconDescription}
        description={hasRenderIcon ? undefined : iconDescription}
        className={`${prefix}--btn__icon`}
        aria-hidden={true}
      />
    );

    let component = 'button';
    let otherProps = {
      disabled,
      type,
    };
    const anchorProps = {
      role: 'button',
      href,
    };
    if (as) {
      component = as;
      otherProps = {
        ...otherProps,
        ...anchorProps,
      };
    } else if (href) {
      component = 'a';
      otherProps = anchorProps;
    }
    return React.createElement(
      component,
      {
        ...other,
        ...commonProps,
        ...otherProps,
      },
      children,
      buttonImage
    );
  }
);

Button.propTypes = {
  /**
   * Specify the content of your Button
   */
  children: PropTypes.node,

  /**
   * Specify how the button itself should be rendered.
   * Make sure to apply all props to the root node and render children appropriately
   */
  as: PropTypes.oneOfType([PropTypes.func, PropTypes.string]),

  /**
   * Specify an optional className to be added to your Button
   */
  className: PropTypes.string,

  /**
   * Specify whether the Button should be disabled, or not
   */
  disabled: PropTypes.bool,

  /**
   * Specify whether the Button should be a small variant
   */
  small: PropTypes.bool,

  /**
   * Specify the kind of Button you want to create
   */
  kind: ButtonTypes.buttonKind.isRequired,

  /**
   * Optionally specify an href for your Button to become an <a> element
   */
  href: PropTypes.string,

  /**
   * Optional prop to specify the tabIndex of the Button
   */
  tabIndex: PropTypes.number,

  /**
   * Optional prop to specify the type of the Button
   */
  type: PropTypes.oneOf(['button', 'reset', 'submit']),

  /**
   * Optional prop to specify the role of the Button
   */
  role: PropTypes.string,

  /**
   * Optional prop to allow overriding the icon rendering.
   * Can be a React component class
   */
  renderIcon: PropTypes.oneOfType([PropTypes.func, PropTypes.object]),

  /**
   * Specify an icon to include in the Button through a string or object
   * representing the SVG data of the icon
   */
  icon: PropTypes.oneOfType([
    PropTypes.shape({
      width: PropTypes.string,
      height: PropTypes.string,
      viewBox: PropTypes.string.isRequired,
      svgData: PropTypes.object.isRequired,
    }),
    PropTypes.string,
    PropTypes.node,
  ]),

  /**
   * If specifying the `icon` prop, provide a description for that icon that can
   * be read by screen readers
   */
  iconDescription: props => {
    if ((props.icon || props.renderIcon) && !props.iconDescription) {
      return new Error(
        'icon/renderIcon property specified without also providing an iconDescription property.'
      );
    }
    return undefined;
  },
};

Button.defaultProps = {
  iconDescription: 'Provide icon description if icon is used',
  tabIndex: 0,
  type: 'button',
  disabled: false,
  small: false,
  kind: 'primary',
};

export default Button;
