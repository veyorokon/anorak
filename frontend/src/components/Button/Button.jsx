import React from "react";
// nodejs library that concatenates classes
import classNames from "classnames";
// nodejs library to set properties for components
import PropTypes from "prop-types";

// material-ui components
import withStyles from "@material-ui/core/styles/withStyles";
import Button from "@material-ui/core/Button";

import buttonStyle from "assets/jss/components/styles/buttonStyle.jsx";

function RegularButton({ ...props }) {
  const {
    classes,
    children,
    disabled,
    secondary,
    color,
    shadow,
    link,
    icon,
    iconPosition,

    round,
    simple,
    size,
    block,
    justIcon,
    className,
    muiClasses,
    loading,
    action,
    ...rest
  } = props;

  const btnClasses = classNames({
    [classes.button]: true,
    [classes[size]]: size,
    [classes[color]]: color,
    [classes.round]: round,
    [classes.secondary]: secondary,
    [classes.loading]: loading,
    [classes.disabled]: disabled,
    [classes.shadow]: shadow,
    [classes.link]: link,

    [classes.simple]: simple,
    [classes.block]: block,
    [classes.link]: link,
    [classes.justIcon]: justIcon,
    [classes.cardAction]: action,
    [className]: className
  });

  if (disabled && icon == null) {
    return (
      <span className={classes.disabledWrapper}>
        <Button {...rest} classes={muiClasses} className={btnClasses}>
          {children}
        </Button>
      </span>
    );
  }
  if (icon) {
    return (
      <Button {...rest} classes={muiClasses} className={btnClasses}>
        <span className={[classes[iconPosition + "Icon"]]}>{icon}</span>
        {children}
      </Button>
    );
  }

  return (
    <Button {...rest} classes={muiClasses} className={btnClasses}>
      {loading ? (
        <div className={classes.loadingDots}>
          <div />
          <div />
          <div />
        </div>
      ) : (
        [children]
      )}
    </Button>
  );
}

RegularButton.propTypes = {
  classes: PropTypes.object.isRequired,
  color: PropTypes.oneOf(["success", "warning", "highlight"]),
  size: PropTypes.oneOf(["sm", "lg"]),
  secondary: PropTypes.bool,
  simple: PropTypes.bool,
  loading: PropTypes.bool,
  shadow: PropTypes.bool,
  link: PropTypes.bool,
  icon: PropTypes.node,
  iconPosition: PropTypes.oneOf(["left", "right"]),

  round: PropTypes.bool,
  disabled: PropTypes.bool,
  block: PropTypes.bool,
  justIcon: PropTypes.bool,
  action: PropTypes.bool,
  className: PropTypes.string,
  // use this to pass the classes props from Material-UI
  muiClasses: PropTypes.object
};

export default withStyles(buttonStyle)(RegularButton);
