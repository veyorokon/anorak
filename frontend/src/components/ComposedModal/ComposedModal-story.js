/**
 * Copyright IBM Corp. 2016, 2018
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { withKnobs, boolean, text } from '@storybook/addon-knobs';
import ComposedModal, {
  ModalHeader,
  ModalBody,
  ModalFooter,
} from '../ComposedModal';
import Button from '../Button';

const props = {
  composedModal: (includeOpen = true) => ({
    open: includeOpen ? boolean('Open (open in <ComposedModal>)', true) : null,
    onKeyDown: action('onKeyDown'),
    danger: boolean('Danger mode (danger)', false),
    selectorPrimaryFocus: text(
      'Primary focus element selector (selectorPrimaryFocus)',
      '[data-modal-primary-focus]'
    ),
  }),
  modalHeader: () => ({
    label: text('Optional Label (label in <ModalHeader>)', 'Optional Label'),
    title: text('Optional title (title in <ModalHeader>)', 'Example'),
    iconDescription: text(
      'Close icon description (iconDescription in <ModalHeader>)',
      'Close'
    ),
    buttonOnClick: action('buttonOnClick'),
  }),
  modalFooter: () => ({
    primaryButtonText: text(
      'Primary button text (primaryButtonText in <ModalFooter>)',
      'Save'
    ),
    primaryButtonDisabled: boolean(
      'Primary button disabled (primaryButtonDisabled in <ModalFooter>)',
      false
    ),
    secondaryButtonText: text(
      'Secondary button text (secondaryButtonText in <ModalFooter>)',
      ''
    ),
    onRequestClose: action('onRequestClose'),
    onRequestSubmit: action('onRequestSubmit'),
  }),
};

storiesOf('ComposedModal', module)
  .addDecorator(withKnobs)
  .add(
    'Using Header / Footer Props',
    () => (
      <ComposedModal {...props.composedModal()}>
        <ModalHeader {...props.modalHeader()} />
        <ModalBody>
          <p className="bx--modal-content__text">
            Please see ModalWrapper for more examples and demo of the
            functionality.
          </p>
        </ModalBody>
        <ModalFooter {...props.modalFooter()} />
      </ComposedModal>
    ),
    {
      info: {
        text: `
            Composed Modal allows you to create your own modal with just the parts you need. The ComposedModal element provides the state management for open/close, as well as passes the ModalHeader a prop to close the modal (with the close button).

            The interior components - ModalHeader / ModalBody / ModalFooter - are all container elements that will render any children you add in, wrapped in the appropriate CSS classes.

            The Modal Header / Modal Footer come with some built in props to let you accelerate towards standard Carbon modal UI. If there are customizations you need to do, see the next example of just using the interior components as containers.
          `,
      },
    }
  )
  .add(
    'Using child nodes',
    () => (
      <ComposedModal {...props.composedModal()}>
        <ModalHeader {...props.modalHeader()}>
          <h1>Testing</h1>
        </ModalHeader>
        <ModalBody>
          <p>
            Please see ModalWrapper for more examples and demo of the
            functionality.
          </p>
        </ModalBody>
        <ModalFooter>
          <Button
            kind={props.composedModal().danger ? 'tertiary' : 'secondary'}>
            Cancel
          </Button>
          <Button
            kind={props.composedModal().danger ? 'danger--primary' : 'primary'}>
            Save
          </Button>
        </ModalFooter>
      </ComposedModal>
    ),
    {
      info: {
        text: `
            Alternatively, you can just use the Modal components as wrapper elements and figure the children out yourself. We do suggest for the header you utilize the built in props for label and title though, for the footer it's mostly a composed element so creating the two buttons yourself (using the Button component) is probably the most straight-forward pattern.
          `,
      },
    }
  )
  .add(
    'Example usage with trigger button',
    () => {
      class ComposedModalExample extends React.Component {
        state = { open: false };
        toggleModal = open => this.setState({ open });
        render() {
          const { open } = this.state;
          return (
            <>
              <Button onClick={() => this.toggleModal(true)}>
                Launch composed modal
              </Button>
              <ComposedModal
                {...props.composedModal()}
                open={open}
                onClose={() => this.toggleModal(false)}>
                <ModalHeader {...props.modalHeader()} />
                <ModalBody>
                  <p className="bx--modal-content__text">
                    Please see ModalWrapper for more examples and demo of the
                    functionality.
                  </p>
                </ModalBody>
                <ModalFooter {...props.modalFooter()} />
              </ComposedModal>
            </>
          );
        }
      }
      return <ComposedModalExample />;
    },
    {
      info: {
        text: `
            An example ComposedModal with a trigger button
          `,
      },
    }
  );
