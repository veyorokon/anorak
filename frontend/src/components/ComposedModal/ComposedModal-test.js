/**
 * Copyright IBM Corp. 2016, 2018
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import { shallow, mount } from 'enzyme';
import Button from '../Button';
import ComposedModal, {
  ModalHeader,
  ModalBody,
  ModalFooter,
} from '../ComposedModal';

describe('<ModalHeader />', () => {
  describe('Renders as expected', () => {
    const titleWrapper = shallow(<ModalHeader title="Something" />);
    const labelWrapper = shallow(<ModalHeader label="Something" />);

    it('does not render title if no title', () => {
      expect(labelWrapper.find('.bx--modal-header__heading').exists()).toBe(
        false
      );
    });

    it('does not render label if no label', () => {
      expect(titleWrapper.find('.bx--modal-header__label').exists()).toBe(
        false
      );
    });

    it('renders title if title text', () => {
      expect(titleWrapper.find('.bx--modal-header__heading').exists()).toBe(
        true
      );
    });

    it('renders label if label text', () => {
      expect(labelWrapper.find('.bx--modal-header__label').exists()).toBe(true);
    });
  });
});

describe('<ModalBody />', () => {
  describe('Renders as expected', () => {
    const wrapper = shallow(
      <ModalBody className="extra-class">
        <p>Test</p>
      </ModalBody>
    );

    it('renders children as expected', () => {
      expect(wrapper.find('p').length).toBe(1);
    });

    it('renders wrapper as expected', () => {
      expect(wrapper.length).toBe(1);
    });

    it('has the expected classes', () => {
      expect(wrapper.hasClass('bx--modal-content')).toEqual(true);
    });

    it('renders extra classes passed in via className', () => {
      expect(wrapper.hasClass('extra-class')).toEqual(true);
    });
  });
});

describe('<ModalFooter />', () => {
  describe('Renders as expected', () => {
    const wrapper = shallow(
      <ModalFooter className="extra-class">
        <p>Test</p>
      </ModalFooter>
    );

    it('renders children as expected', () => {
      expect(wrapper.find('p').length).toBe(1);
    });

    it('renders wrapper as expected', () => {
      expect(wrapper.length).toBe(1);
    });

    it('has the expected classes', () => {
      expect(wrapper.hasClass('bx--modal-footer')).toEqual(true);
    });

    it('renders extra classes passed in via className', () => {
      expect(wrapper.hasClass('extra-class')).toEqual(true);
    });
  });

  describe('Should render buttons only if appropriate prop passed in in', () => {
    const wrapper = shallow(
      <ModalFooter className="extra-class">
        <p>Test</p>
      </ModalFooter>
    );

    const primaryWrapper = shallow(<ModalFooter primaryButtonText="test" />);
    const secondaryWrapper = shallow(
      <ModalFooter secondaryButtonText="test" />
    );

    it('does not render primary button if no primary text', () => {
      expect(wrapper.find('.bx--btn--primary').exists()).toBe(false);
    });

    it('does not render secondary button if no secondary text', () => {
      expect(wrapper.find('.bx--btn--secondary').exists()).toBe(false);
    });

    it('renders primary button if primary text', () => {
      const buttonComponent = primaryWrapper.find(Button);
      expect(buttonComponent.exists()).toBe(true);
      expect(buttonComponent.props().kind).toBe('primary');
    });

    it('renders primary button if secondary text', () => {
      const buttonComponent = secondaryWrapper.find(Button);
      expect(buttonComponent.exists()).toBe(true);
      expect(buttonComponent.props().kind).toBe('secondary');
    });
  });

  describe('Should render the appropriate buttons when `danger` prop is true', () => {
    const primaryWrapper = shallow(
      <ModalFooter primaryButtonText="test" danger />
    );
    const secondaryWrapper = shallow(
      <ModalFooter secondaryButtonText="test" danger />
    );

    it('renders danger--primary button if primary text && danger', () => {
      const buttonComponent = primaryWrapper.find(Button);
      expect(buttonComponent.exists()).toBe(true);
      expect(buttonComponent.props().kind).toBe('danger--primary');
    });

    it('renders tertiary button if secondary text && danger', () => {
      const buttonComponent = secondaryWrapper.find(Button);
      expect(buttonComponent.exists()).toBe(true);
      expect(buttonComponent.props().kind).toBe('tertiary');
    });
  });
});

describe('<ComposedModal />', () => {
  it('renders', () => {
    const wrapper = mount(<ComposedModal open />);
    expect(wrapper).toMatchSnapshot();
  });

  it('changes the open state upon change in props', () => {
    const wrapper = mount(<ComposedModal open />);
    expect(wrapper.state().open).toEqual(true);
    wrapper.setProps({ open: false });
    expect(wrapper.state().open).toEqual(false);
  });

  it('avoids change the open state upon setting props, unless there the value actually changes', () => {
    const wrapper = shallow(<ComposedModal />);
    wrapper.setProps({ open: true });
    wrapper.setState({ open: false });
    wrapper.setProps({ open: true });
    expect(wrapper.state().open).toEqual(false);
  });

  it('calls onClick upon user-initiated closing', () => {
    const onClose = jest.fn();
    const wrapper = mount(
      <ComposedModal open onClose={onClose}>
        <ModalHeader />
      </ComposedModal>
    );
    const button = wrapper.find('.bx--modal-close').first();
    button.simulate('click');
    expect(wrapper.state().open).toEqual(false);
    expect(onClose.mock.calls.length).toBe(1);
  });

  it('provides a way to prevent upon user-initiated closing', () => {
    const onClose = jest.fn(() => false);
    const wrapper = mount(
      <ComposedModal open onClose={onClose}>
        <ModalHeader />
      </ComposedModal>
    );
    const button = wrapper.find('.bx--modal-close').first();
    button.simulate('click');
    expect(wrapper.state().open).toEqual(true);
  });

  it('should focus on the primary actionable button in ModalFooter by default', () => {
    mount(
      <ComposedModal open>
        <ModalFooter primaryButtonText="Save" />
      </ComposedModal>
    );
    expect(
      document.activeElement.classList.contains('bx--btn--primary')
    ).toEqual(true);
  });

  it('should focus on the element that matches selectorPrimaryFocus', () => {
    mount(
      <ComposedModal open selectorPrimaryFocus=".bx--modal-close">
        <ModalHeader label="Optional Label" title="Example" />
        <ModalFooter primaryButtonText="Save" />
      </ComposedModal>
    );
    expect(
      document.activeElement.classList.contains('bx--modal-close')
    ).toEqual(true);
  });
});
