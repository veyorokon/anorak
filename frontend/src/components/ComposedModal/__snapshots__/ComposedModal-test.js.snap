// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ComposedModal /> renders 1`] = `
<ComposedModal
  onKeyDown={[Function]}
  open={true}
  selectorPrimaryFocus="[data-modal-primary-focus]"
>
  <div
    className="bx--modal is-visible"
    onBlur={[Function]}
    onClick={[Function]}
    onKeyDown={[Function]}
    onTransitionEnd={[Function]}
    open={true}
    role="presentation"
    tabIndex={-1}
  >
    <div
      className="bx--modal-container"
    />
  </div>
</ComposedModal>
`;
